pkg/admission/http.go

```diff
@@ -0,0 +1,90 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package admission
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+
+	"k8s.io/api/admission/v1beta1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+)
+
+var (
+	scheme = runtime.NewScheme()
+	codecs = serializer.NewCodecFactory(scheme)
+)
+
+type httpHandler struct {
+	*Webhook
+	Annotations map[string]string
+}
+
+func (h httpHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	var body []byte
+	var err error
+
+	var reviewResponse *Response
+	if r.Body != nil {
+		if body, err = ioutil.ReadAll(r.Body); err != nil {
+			log.Error(err, "unable to read the body from the incoming request")
+			reviewResponse = ErrorResponse(http.StatusBadRequest, err)
+			marshalAndWrite(w, reviewResponse)
+			return
+		}
+	}
+
+	// verify the content type is accurate
+	contentType := r.Header.Get("Content-Type")
+	if contentType != "application/json" {
+		err = fmt.Errorf("contentType=%s, expect application/json", contentType)
+		log.Error(err, "unable to process a request with an unknown content type")
+		reviewResponse = ErrorResponse(http.StatusBadRequest, err)
+		marshalAndWrite(w, reviewResponse)
+		return
+	}
+
+	ar := v1beta1.AdmissionReview{}
+	deserializer := codecs.UniversalDeserializer()
+	if _, _, err := deserializer.Decode(body, nil, &ar); err != nil {
+		log.Error(err, "unable to decode the request")
+		reviewResponse = ErrorResponse(http.StatusBadRequest, err)
+		marshalAndWrite(w, reviewResponse)
+		return
+	}
+
+	response := v1beta1.AdmissionReview{}
+	if reviewResponse != nil {
+		resp := h.Handle(&Request{Request: ar.Request, Annotations: h.Annotations})
```

Comment:
Not a blocker, but we should think about the cases where one of the registered handler panics, we should have panic-recovery somewhere here.